app/models/assignment.rb

```diff
@@ -272,15 +272,26 @@ def marking_started?
           .any?
   end
 
+  def results_fails
```

Comment:
One of the downsides of splitting up the previous computation into two methods is the call to `Result.student_marks_by_assignment` is repeated, which isn't great.

Take a look at how `GradeEntryForm` does it: it uses a helper `percentage_grades_array`, which actually saves the grades array in an instance variable (`@grades_array`) so that's it's saved across the two calls. You can use the same approach for these two methods, which hopefully will allow you to:

1. Pull the implementation of these two methods up to the parent `Assessment` class.
2. Remove `Result.student_marks_by_assignment` altogether.